### PR TITLE
indexer: fix missing tx/block indexing

### LINKED ISSUE #7312: tx indexer miss block when restart
**Tendermint v0.34.14**

**ABCI app**

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:

We found that our node miss some tx indexes for certain blocks.
When investigating the code, we found a case that when you stop the node while the tx indexer is indexing a block, that block won't be picked up again after the restart, and it won't be indexed ever.

**What you expected to happen**:

tx indexer should index all the blocks.

**Have you tried the latest version**: no

**How to reproduce it** (as minimally and precisely as possible):

1. Patch `v0.34.14` like this for easier reproduction.

```diff
--- a/state/txindex/indexer_service.go
+++ b/state/txindex/indexer_service.go
@@ -2,6 +2,7 @@ package txindex

 import (
        "context"
+       "time"

        "github.com/tendermint/tendermint/libs/service"
        "github.com/tendermint/tendermint/state/indexer"
@@ -61,6 +62,7 @@ func (is *IndexerService) OnStart() error {
                        eventDataHeader := msg.Data().(types.EventDataNewBlockHeader)
                        height := eventDataHeader.Header.Height
                        batch := NewBatch(eventDataHeader.NumTxs)
+                       is.Logger.Error("[debug] start index block", "height", height)

                        for i := int64(0); i < eventDataHeader.NumTxs; i++ {
                                msg2 := <-txsSub.Out()
@@ -82,11 +84,14 @@ func (is *IndexerService) OnStart() error {
                                is.Logger.Info("indexed block", "height", height)
                        }

+                       time.Sleep(time.Second)
                        if err = is.txIdxr.AddBatch(batch); err != nil {
                                is.Logger.Error("failed to index block txs", "height", height, "err", err)
                        } else {
                                is.Logger.Debug("indexed block txs", "height", height, "num_txs", eventDataHeader.NumTxs)
                        }
+
+                       is.Logger.Error("[debug] finish index block", "height", height)
                }
        }()
        return nil
```

2. Start devnet, and randomly restart the node.
3. You can observe logs similar to this:
```
...
2:16PM ERR [debug] start index block height=62 module=txindex server=node
2:16PM ERR [debug] start index block height=63 module=txindex server=node
2:16PM ERR [debug] finish index block height=63 module=txindex server=node
...
```

It means block 62 is not indexed.

### PR DESCRIPTION
**Pull request by JayT106**
The indexer might not finish the indexing when terminating the node.
1. Handle blockHeadersSub.Cancelled event in the goroutine of the indexer service.
2. Re-index the block/tx events when replaying the final block. 

Closes #7312

### CODE REVIEW

File: state/txindex/indexer_service.go [hunk 1]
```diff
@@ -55,37 +57,45 @@ func (is *IndexerService) OnStart() error {
 		return err
 	}
 
+	is.wg = &sync.WaitGroup{}
```
Comment (creachadair) [inline state/txindex/indexer_service.go:60]: Adding a goroutine here does not appear related to the main purpose of this PR. Let's please leave this out, if there's a good reason for it we can do it in a separate PR.

File: test/maverick/consensus/replay.go [hunk 2]
```diff
@@ -469,7 +469,16 @@ func (h *Handshaker) replayBlocks(
 			assertAppHashEqualsOneFromBlock(appHash, block)
 		}
 
-		appHash, err = sm.ExecCommitBlock(proxyApp.Consensus(), block, h.logger, h.stateStore, h.genDoc.InitialHeight)
+		appHash, err = sm.ExecCommitBlock(
```
Comment (creachadair) [inline test/maverick/consensus/replay.go:472]: While the overall goal here is fine, adding random extra parameters to an existing exported API doesn't seem like a good choice—especially since the new arguments are all to serve a special case. I especially don't like the naked Boolean; that pattern has been a real hassle for ongoing maintenance in my experience.

I think a better approach here would be to add a new method, perhaps taking an argument struct. We can re-implement the existing method in terms of that, to avoid duplication, but at least then the usage will be more self-documenting.

**Review by creachadair — CHANGES_REQUESTED** [review]

File: test/maverick/consensus/replay.go (hunk 2 above)
Comment (JayT106) [reply test/maverick/consensus/replay.go:472]: sure, will redo it.

File: state/txindex/indexer_service.go (hunk 1 above)
Comment (JayT106) [reply state/txindex/indexer_service.go]: removed

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -26,3 +26,5 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 
 - [\#7057](https://github.com/tendermint/tendermint/pull/7057) Import Postgres driver support for the psql indexer (@creachadair).
 - [\#7106](https://github.com/tendermint/tendermint/pull/7106) Revert mutex change to ABCI Clients (@tychoish).
+
```
Comment (tychoish) [inline CHANGELOG_PENDING.md:29]: omit this line.

File: state/execution.go [hunk 4]
```diff
@@ -50,6 +50,18 @@ func BlockExecutorWithMetrics(metrics *Metrics) BlockExecutorOption {
 	}
 }
 
+// BlockCommitExecParams provides context for executing the block commit.
+type BlockCommitExecParams struct {
```
Comment (tychoish) [inline state/execution.go:54]: I'm a little worried that we're making a struct of a lot of things that are pointers. I don't know if there's anything better to be done, but... uff.

**Review by tychoish — COMMENTED** [review]
I think this seems reasonable, but i'm a touch worried by the fact that this goes in only to 0.34.x? Why wouldn't this be a problem in mainline?

**Comment by JayT106** [comment]
> I think this seems reasonable, but i'm a touch worried by the fact that this goes in only to 0.34.x? Why wouldn't this be a problem in mainline?

@creachadair had fixed the issue in mainline and possible to BP to 0.35.x, but the change is not feasible for 0.34.x
https://github.com/tendermint/tendermint/issues/7312#issuecomment-978427997

File: state/execution.go (hunk 4 above)
Comment (JayT106) [reply state/execution.go:54]: removed the pointer of the validator.
for the block pointer, it requires bigger changes, should I do it in this PR?

**Comment by creachadair** [comment]
> > I think this seems reasonable, but i'm a touch worried by the fact that this goes in only to 0.34.x? Why wouldn't this be a problem in mainline?
> 
> @creachadair had fixed the issue in mainline and possible to BP to 0.35.x, but the change is not feasible for 0.34.x [#7312 (comment)](https://github.com/tendermint/tendermint/issues/7312#issuecomment-978427997)

I'm not sure if the changes I made at tip actually fix this issue—I didn't change which events get logged, only how they are delivered to the indexer. It looks like this change maybe also generates events during replay that maybe we weren't before?

**Comment by JayT106** [comment]
> > > I think this seems reasonable, but i'm a touch worried by the fact that this goes in only to 0.34.x? Why wouldn't this be a problem in mainline?
> > 
> > 
> > @creachadair had fixed the issue in mainline and possible to BP to 0.35.x, but the change is not feasible for 0.34.x [#7312 (comment)](https://github.com/tendermint/tendermint/issues/7312#issuecomment-978427997)
> 
> I'm not sure if the changes I made at tip actually fix this issue—I didn't change which events get logged, only how they are delivered to the indexer. It looks like this change maybe also generates events during replay that maybe we weren't before?

I think so, the current mainline behavior will index the missing events during the replay.
